### PR TITLE
Add RDS and CDS

### PR DESCRIPTION
**Pull request by kyessenov**
Large change due to code shuffle from agent to discovery service.
Added ProxyContext to capture local proxy information context.
Removed envoy epoch checks, since envoy doesn't restart anymore for HTTP rules sometimes.
Added golden tests for the discovery service.

### CODE REVIEW

File: proxy/envoy/discovery.go [hunk 1]
```diff
@@ -87,27 +109,99 @@ func (ds *DiscoveryService) Run() {
...
+	}
+
+	if err := response.WriteEntity(ClusterManager{Clusters: clusters}); err != nil {
+		glog.Warning(err)
+	}
+}
+
+// ListRoutes responds to RDS requests, used by HTTP routes
+func (ds *DiscoveryService) ListRoutes(request *restful.Request, response *restful.Response) {
+	if serviceCluster := request.PathParameter(ServiceCluster); serviceCluster != IstioServiceCluster {
```
Comment (rshriram) [inline proxy/envoy/discovery.go]: This is effectively disabling all source based routing isn't it? Why can't we run the RDS and CDS on localhost next to Envoy itself so that the proxy agents know their parent services? This is where the explicit service affinity declaration comes in.

File: proxy/envoy/testdata/envoy-cb-policy.json.golden [hunk 2]
```diff
@@ -311,74 +214,22 @@
...
-            "max_requests": 100
-          }
-        },
-        "outlier_detection": {
-          "consecutive_5xx": 10,
-          "interval_ms": 30000,
-          "base_ejection_time_ms": 15500,
-          "max_ejection_percent": 100
-        }
+        "lb_type": "round_robin"
```
Comment (rshriram) [inline proxy/envoy/testdata/envoy-cb-policy.json.golden]: Why did you remove the circuit breakers ?

File: proxy/envoy/testdata/envoy-timeout-rule.json.golden [hunk 3]
```diff
@@ -9,62 +9,10 @@
...
-                    "10.2.0.0:80",
-                    "10.2.0.0"
-                  ],
-                  "routes": [
-                    {
-                      "prefix": "/",
-                      "cluster": "outbound:world.default.svc.cluster.local:http",
-                      "timeout_ms": 30000,
-                      "retry_policy": {
-                        "retry_on": "5xx,connect-failure,refused-stream",
```
Comment (rshriram) [inline proxy/envoy/testdata/envoy-timeout-rule.json.golden]: And these as well

File: proxy/envoy/discovery.go (hunk 1 above)
Comment (kyessenov) [inline proxy/envoy/discovery.go]: Proxies send `service-node` which is pod IP. Discovery computes routes for that IP (by using `HostInstances` call to the service registry). Source-based routing is already being tested by the end-to-end test, although no unit tests. I can add unit tests for more coverage for calls to `ListRoutes`.

File: proxy/envoy/testdata/envoy-cb-policy.json.golden (hunk 2 above)
Comment (kyessenov) [inline proxy/envoy/testdata/envoy-cb-policy.json.golden]: These are now part of CDS output. I need to add a unit test with the sample output from CDS.

File: proxy/envoy/testdata/envoy-timeout-rule.json.golden (hunk 3 above)
Comment (kyessenov) [inline proxy/envoy/testdata/envoy-timeout-rule.json.golden]: This is now output of RDS. It turns out envoy config is identical for all three cases here, and we should be using RDS/CDS output instead.

File: proxy/envoy/config.go [hunk 4]
```diff
@@ -102,59 +102,35 @@ func Generate(instances []*model.ServiceInstance, services []*model.Service,
 		},
 		ClusterManager: ClusterManager{
 			Clusters: clusters,
-			SDS: SDS{
-				Cluster:        buildSDSCluster(mesh),
+			SDS: &SDS{
+				Cluster:        buildDiscoveryCluster(mesh.DiscoveryAddress, "sds"),
+				RefreshDelayMs: 1000,
```
Comment (rshriram) [inline proxy/envoy/config.go]: Please add a TODO note here to make these three fields configurable (sds, cds, rds)

File: proxy/envoy/agent.go [hunk 5]
```diff
@@ -34,9 +34,11 @@ type agent struct {
...
 	// Envoy config root
 	configRoot string
-	// serviceCluster is the first component of the local proxy identityi (cluster name)
+	// serviceCluster is the first component of the local proxy identity
+	// (cluster name, e.g. "istio-proxy")
 	serviceCluster string
-	// serviceNode is the second component of the local proxy identity (node name)
+	// serviceNode is the second component of the local proxy identity
+	// (node name, e.g. pod name)
 	serviceNode string
```
Comment (rshriram) [inline proxy/envoy/agent.go]: please add a note here about how source based routing is done.. 
"we use pod IP to identify the source service that the pod belongs to and filter route rules accordingly" or whatever you see fit.

File: proxy/envoy/discovery.go (hunk 1 above)
Comment (rshriram) [inline proxy/envoy/discovery.go]: Can you please add this comment to the function ListRoutes ?
